src/expr/src/relation/func.rs

```diff
@@ -390,40 +398,69 @@ fn row_number<'a, I>(
     callers_temp_storage: &'a RowArena,
     order_by: &[ColumnOrder],
 ) -> Datum<'a>
+where
+    I: IntoIterator<Item = Datum<'a>>,
+{
+    // We want to use our own temp_storage here, to avoid flooding `callers_temp_storage` with a
+    // large number of new datums. This is because we don't want to make an assumption about
+    // whether the caller creates a new temp_storage between window partitions.
+    let temp_storage = RowArena::new();
+    let datums = row_number_no_list(datums, &temp_storage, order_by);
+
+    callers_temp_storage.make_datum(|packer| {
+        packer.push_list(datums);
+    })
+}
+
+fn row_number_no_list<'a: 'b, 'b, I>(
+    datums: I,
+    callers_temp_storage: &'b RowArena,
+    order_by: &[ColumnOrder],
+) -> impl Iterator<Item = Datum<'b>>
 where
     I: IntoIterator<Item = Datum<'a>>,
 {
     let datums = order_aggregate_datums(datums, order_by);
 
-    let temp_storage = RowArena::with_capacity(datums.size_hint().0);
-    let datums = datums
+    callers_temp_storage.reserve(datums.size_hint().0);
```

Comment:
Why do we switch to `callers_temp_storage` here?

Reply:
The problem is that this arena will need to live after returning from this function, because the `make_datum` call is in the closure that is part of constructing the `Iterator` that is returned from this function. (At first glance, it seems that we could just add `move` to the `map`'s closure, but this doesn't work, because it's an `FnMut` closure, and the reference would escape through the result of `make_datum` being returned.)